### PR TITLE
Build autograding process

### PR DESCRIPTION
**Pull request by ARaps1**
As a system we want to be able to auto grade a users assignments and provide the results back to the system.

Acceptance Criteria

- [x]  Setup a basic action using the documentation [[here](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action)](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action)
- [x]  Modify the action to use typescript and copy over our eslint/prettier configs from the main repo
- [x]  Update the Action to perform the following steps:
    - [x]  Generate an OIDC id token using the id token method available in `@actions/core`
    - [x]  Call our [[Github Actions Auth](https://www.notion.so/Github-Actions-Auth-3234cd5f3a10448482e4c62e86af8bdf)](https://www.notion.so/Github-Actions-Auth-3234cd5f3a10448482e4c62e86af8bdf) endpoint with the token to generate a Matric token
    - [x]  Call our [[Retrieve Github repo as autograder](https://www.notion.so/Retrieve-Github-repo-as-autograder-561985bf0b8c4875b31c419272dc8548)](https://www.notion.so/Retrieve-Github-repo-as-autograder-561985bf0b8c4875b31c419272dc8548) endpoint with the new token and the name of the repo
    - [x]  Git clone the url returned from the api.
    - [x]  Rename the path of the cloned repo to **`source`**
    - [x]  Rename the path to the the user repo to **`submission`**
    - [x]  Create a results folder
    - [x]  Copy the [[setup.sh](http://setup.sh/)](http://setup.sh) file (if it exists) from the source folder up to the top folder
    - [x]  execute the [[setup.sh](http://setup.sh/)](http://setup.sh) file (if it exists) and redirect the output to a file
    - [x]  Copy the run_autograder file from the source folder up to the top folder
    - [x]  execute the run_autograder file and redirect the output to a file
    - [x]  Verify that there is a file in **`results/results.json`**
    - [x]  Read the file into memory and validate that it is json and it matches [[the expected schema](https://www.notion.so/Accept-submission-from-the-autograder-b224719e18534789a0eb3ba00abd7d0b)](https://www.notion.so/Accept-submission-from-the-autograder-b224719e18534789a0eb3ba00abd7d0b)
    - [x]  Send the json along with some metadata of the request (commit id, repo name, actor, etc) to the [[Accept submission from the autograder](https://www.notion.so/Accept-submission-from-the-autograder-b224719e18534789a0eb3ba00abd7d0b)](https://www.notion.so/Accept-submission-from-the-autograder-b224719e18534789a0eb3ba00abd7d0b) endpoint

### CODE REVIEW

**Comment by ARaps1** [comment]
Clone Public Repository tests will fail as along as we don't have our backend deployed.

**Review by JJamesWWang — COMMENTED** [review]
~~Looks like you accidentally committed `dist/index.js`~~
> when the action is called from other repos it looks at the dist/index.js
> https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action

File: action.yml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+name: 'Autograder'
+description: 'Autograde a repo submission and return results'
+runs:
+  using: 'node16'
```
Comment (JJamesWWang) [inline action.yml:4]: Should this be node18?

File: src/index.ts [hunk 2]
```diff
@@ -1 +1,166 @@
-console.log('hello world!');
+require('dotenv').config();
```
Comment (JJamesWWang) [inline src/index.ts]: Is this fixable?

File: src/util/autograder.interface.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
+export interface AutograderData {
+  //What an autograder returns
+  score: number; // optional, but required if not on each test case below. Overrides total of tests if specified.
+  execution_time: number; // optional, seconds
+  output: string; // optional
+  visibility: Visibility; // Optional visibility setting
+  stdout_visibility: Visibility; // Optional stdout visibility setting
+  extra_data: any; // Optional extra data to be stored
```
Comment (JJamesWWang) [inline src/util/autograder.interface.ts]: Is it possible to make it a union of types? Or are there an infinite number of structures that this data can take

File: src/util/file-util.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+): Promise<boolean> {
+  if (!fileExists(oldPath)) {
+    console.log(`${oldPath} does not exists!`);
+    return false;
+  }
+
+  return new Promise((resolve, reject) => {
+    exec(
+      `${cmd} ${oldPath} ${newPath}`,
+      (error: any, _stdout: any, _stderr: any) => {
```
Comment (JJamesWWang) [inline src/util/file-util.ts]: Same here as autograder.interface.ts

File: src/util/file-util.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        if (error != null) {
```
Comment (JJamesWWang) [inline src/util/file-util.ts:19]: It's weird that it's complaining about stdout and stderr even though you have an underscore... maybe you can remove them entirely?

File: src/util/file-util.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        return;
+      }
+      const wroteSuccessfully = await writeOutputToFile(stdout, outputFilePath);
+      resolve(wroteSuccessfully);
+    });
+  });
+}
+
+export async function createFolder(path: string): Promise<void> {
+  exec(`mkdir ${path}`, (_error: any, _stdout: any, _stderr: any) => {});
```
Comment (JJamesWWang) [inline src/util/file-util.ts]: See if you can leave out the second arg here too

File: src/util/json-util.ts [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+  if (!Array.isArray(data)) {
+    return false;
+  }
+
+  for (const obj of data) {
+    if (typeof obj != 'object') {
+      return false;
+    }
+
+    if (!propertyExists('name', obj) || !obj.hasOwnProperty('value')) {
```
Comment (JJamesWWang) [inline src/util/json-util.ts]: idk what this means but how is hasOwnProperty different from propertyExists

**Review by JJamesWWang — CHANGES_REQUESTED** [review]
I'm inclined to ask you to find a way to make the ESLint workflow not fail - it'll be annoying if everyone else's branches break because of these changes and then it becomes hard to tell if it was their own change or these changes

Just realized a lot of the comments below don't show the ESLint error - most of them are in referral to that.

File: src/util/json-util.ts (hunk 7 above)
Comment (ARaps1) [reply src/util/json-util.ts]: This one I have to fix. I missed switching it to the propertyExists

File: action.yml (hunk 1 above)
Comment (ARaps1) [reply action.yml:4]: Let me confirm this

File: src/util/autograder.interface.ts (hunk 3 above)
Comment (ARaps1) [reply src/util/autograder.interface.ts]: There are infinite number of structures you can come up with because of the optionals

**Review by JJamesWWang — APPROVED** [review]
Looks good to me, seems we can't properly verify this until later

**Pull request merged**